net/libfab/libfab.c

```diff
@@ -0,0 +1,3455 @@
+		iov[i].len  = nb->nb_buffer.ov_vec.v_count[i];
+	}
+
+	return M0_RC(0);
+}
+
+/**
+ * Decodes the descriptor of a (passive) network buffer.
+ */
+static void libfab_bdesc_decode(struct m0_fab__buf *fb, 
+				struct m0_fab__ep_name *epname)
+{
+	struct m0_net_buffer *nb = fb->fb_nb;
+	struct m0_fab__ndom  *ndom = nb->nb_dom->nd_xprt_private;
+
+	fb->fb_rbd = (struct m0_fab__bdesc *)(nb->nb_desc.nbd_data);
+	fb->fb_riov = (struct fi_rma_iov *)(nb->nb_desc.nbd_data + 
+					    sizeof(struct m0_fab__bdesc));
+	libfab_ep_ntop(fb->fb_rbd->fbd_netaddr, epname);
+	M0_ASSERT(fb->fb_rbd->fbd_iov_cnt <= ndom->fnd_seg_nr);
+}
+
+/**
+ * Register the buffer with the appropriate access to the domain
+ */
+static int libfab_buf_dom_reg(struct m0_net_buffer *nb, struct fid_domain *dp)
+{
+	struct m0_fab__buf    *fbp;
+	struct m0_fab__buf_mr *mr;
+	struct m0_fab__ndom   *ndom;
+	uint64_t               key;
+	uint32_t               retry_cnt;
+	int                    seg_nr;
+	int                    i;
+	int                    ret = FI_SUCCESS;
+
+	M0_PRE(nb != NULL && nb->nb_dom != NULL);
+	fbp = nb->nb_xprt_private;
+	seg_nr = nb->nb_buffer.ov_vec.v_nr;
+	ndom = nb->nb_dom->nd_xprt_private;
+
+	M0_ASSERT(fbp != NULL && dp != NULL && ndom != NULL);
+	M0_ASSERT(seg_nr <= ndom->fnd_seg_nr);
+
+	mr = &fbp->fb_mr;
+	if (fbp->fb_dp == dp)
+		return M0_RC(ret);
+
+	if (fbp->fb_state == FAB_BUF_REGISTERED)
+		M0_LOG(M0_ERROR,"Re-registeration of buffer");
+
+	for (i = 0; i < seg_nr; i++) {
+		retry_cnt = 20;
+		key = libfab_mr_keygen();
+		
+		ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i],
+				nb->nb_buffer.ov_vec.v_count[i],
+				FAB_MR_ACCESS, FAB_MR_OFFSET, key,
+				FAB_MR_FLAG, &mr->bm_mr[i], NULL);
+
+		if (ret != FI_SUCCESS) {
+			M0_LOG(M0_ERROR, "fi_mr_reg = %d i =%d key = %"PRIu64,
+			       ret, i, key);
+			while (ret != FI_SUCCESS && retry_cnt > 0) {
+				key = libfab_mr_keygen();
+				ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i],
+						nb->nb_buffer.ov_vec.v_count[i],
+						FAB_MR_ACCESS, FAB_MR_OFFSET,
+						key, FAB_MR_FLAG, &mr->bm_mr[i],
+						NULL);
+				retry_cnt--;
+			}
+			M0_ASSERT(ret == FI_SUCCESS);
+		}
+
+		mr->bm_desc[i] = fi_mr_desc(mr->bm_mr[i]);
+	}
+
+	fbp->fb_dp = dp;
+	fbp->fb_state = FAB_BUF_REGISTERED;
+
+	return M0_RC(ret);
+}
+
+/**
+ * Buffers are queued before connection establishment, send those buffers
+ * after connection establishment
+ */
+static void libfab_pending_bufs_send(struct m0_fab__ep *ep)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *fbp;
+	struct m0_net_buffer     *nb;
+	int                       ret = 0;
+
+	aep = libfab_aep_get(ep);
+	m0_tl_for(fab_sndbuf, &ep->fep_sndbuf, fbp) {
+		nb = fbp->fb_nb;
+		fbp->fb_txctx = ep;
+		switch (nb->nb_qtype) {
+			case M0_NET_QT_MSG_SEND:
+			case M0_NET_QT_ACTIVE_BULK_RECV:
+			case M0_NET_QT_ACTIVE_BULK_SEND:
+				ret = libfab_txbuf_list_add(libfab_buf_ma(nb),
+							    fbp, aep);
+				libfab_bufq_process(libfab_buf_ma(nb));
```

Comment:
When we try to post the buffer to the libfabric queue and it fails because the libfabric queue is full, then it will remain on our queue unless the buffer timeout occurs or the buffer is successfully posted.